db/eventlog.c

```diff
@@ -547,34 +534,40 @@ static void eventlog_add_int(cson_object *obj, const struct reqlogger *logger)
     eventlog_path(obj, logger);
 }
 
+static inline void add_to_fingerprints(const struct reqlogger *logger)
+{
+    bool isSql = logger->event_type && (strcmp(logger->event_type, "sql") == 0);
+    bool isSqlErr = logger->error && logger->stmt;
+
+    if ((isSql || isSqlErr) && !hash_find(seen_sql, logger->fingerprint)) {
+        eventlog_add_newsql(logger);
+    }
+}
+
 void eventlog_add(const struct reqlogger *logger)
 {
-    Pthread_mutex_lock(&eventlog_lk);
     if (eventlog == NULL || !eventlog_enabled) {
-        Pthread_mutex_unlock(&eventlog_lk);
         return;
     }
 
-    cson_value *val;
-    cson_object *obj;
-
-    eventlog_count++;
-    if (eventlog_every_n > 1 && eventlog_count % eventlog_every_n != 0) {
-        Pthread_mutex_unlock(&eventlog_lk);
+    int loc_count = ATOMIC_ADD64(eventlog_count, 1);
+    if (eventlog_every_n > 1 && loc_count % eventlog_every_n != 0) {
         return;
     }
-    if (bytes_written > eventlog_rollat) {
-        eventlog_roll();
-    }
-    Pthread_mutex_unlock(&eventlog_lk);
 
-    val = cson_value_new_object();
-    obj = cson_value_get_object(val);
-    eventlog_add_int(obj, logger);
+    cson_value *val = cson_value_new_object();
+    cson_object *obj = cson_value_get_object(val);
+    populate_obj(obj, logger);
 
     Pthread_mutex_lock(&eventlog_lk);
-    if (eventlog != NULL && eventlog_enabled)
+
+    if (eventlog != NULL && eventlog_enabled) {
+        if (bytes_written > eventlog_rollat) {
```

Comment:
This will now roll, and add to fingerprints on top of what it was doing earlier. The time spent with the lock held is possibly similar to before, but at least we are lock/release only once instead of several times, so contention and total time should be less.